### PR TITLE
ppc64le: add 128-bit IEEE long double ABI plumbing

### PR DESCRIPTION
**Pull request by gyakovlev**

### CODE REVIEW

**Comment by gyakovlev** [comment]
rebased and tested

```bash
 * /bin/sh /var/tmp/portage/sys-devel/gcc-11.3.0/work/gcc-11.3.0/configure --host=powerpc64le-unknown-linux-gnu --build=powerpc64le-unknown-linux-gnu --prefix=/usr --bindir=/usr/powerpc64le-unknown-linux-gnu/gcc-bin/11.3.0 --includedir=/usr/lib/gcc/powerpc
64le-unknown-linux-gnu/11.3.0/include --datadir=/usr/share/gcc-data/powerpc64le-unknown-linux-gnu/11.3.0 --mandir=/usr/share/gcc-data/powerpc64le-unknown-linux-gnu/11.3.0/man --infodir=/usr/share/gcc-data/powerpc64le-unknown-linux-gnu/11.3.0/info --with-gx
x-include-dir=/usr/lib/gcc/powerpc64le-unknown-linux-gnu/11.3.0/include/g++-v11 --with-python-dir=/share/gcc-data/powerpc64le-unknown-linux-gnu/11.3.0/python --enable-languages=c,c++,fortran --enable-obsolete --enable-secureplt --disable-werror --with-syst
em-zlib --disable-nls --disable-libunwind-exceptions --enable-checking=release --with-bugurl=https://bugs.gentoo.org/ --with-pkgversion=Gentoo Hardened 11.3.0 p7 --enable-esp --enable-libstdcxx-time --disable-libstdcxx-pch --enable-shared --enable-threads=
posix --enable-__cxa_atexit --enable-clocale=gnu --disable-multilib --disable-fixed-point --with-long-double-format=ieee --enable-targets=all --enable-libgomp --disable-libssp --disable-libada --disable-cet --disable-systemtap --disable-valgrind-annotation
s --disable-vtable-verify --disable-libvtv --with-zstd --enable-lto --with-isl --disable-isl-version-check --enable-default-pie --enable-default-ssp --with-cpu=power9 --with-build-config=bootstrap-lto --disable-libada --disable-libsanitizer --disable-libvt
v --disable-libgomp --disable-libquadmath --disable-libatomic --disable-lto --disable-bootstrap --enable-host-shared --enable-languages=jit    
```

as you see it's passing `--with-long-double-format=ieee` just fine.

**Comment by gyakovlev** [comment]
just for clarity, `--with-cpu=power9` comes from my custom `EXTRA_ECONF`

File: eclass/toolchain.eclass [hunk 1]
```diff
@@ -277,6 +277,7 @@ if [[ ${PN} != kgcc64 && ${PN} != gcc-* ]] ; then
 	tc_version_is_at_least 10 && IUSE+=" zstd" TC_FEATURES+=( zstd )
 	tc_version_is_at_least 11 && IUSE+=" valgrind" TC_FEATURES+=( valgrind )
 	tc_version_is_at_least 11 && IUSE+=" custom-cflags"
+	tc_version_is_at_least 11 && IUSE+=" ieee-long-double"
```
Comment (thesamesam) [inline eclass/toolchain.eclass:280]: I suspect we don't want to do this for 11, only 12+, to narrow the things we're actually testing with.

**Comment by gyakovlev** [comment]
will 12 get stable by 23.0 profile release?

also for profile sake I'd like to use.force it on gcc and mask versions that don't have it, because without it gcc will produce different abi... we want  to guarantee abi uses that.

```bash
cc -E -dM - < /dev/null | grep -P 'LDBL|FLT128' | sort
```
...
```diff
 #define __FLT128_MIN__ 3.36210314311209350626267781732175260e-4932F128
 #define __FLT128_MIN_EXP__ (-16381)
 #define __FLT128_NORM_MAX__ 1.18973149535723176508575932662800702e+4932F128
-#define __LDBL_DECIMAL_DIG__ 33
-#define __LDBL_DENORM_MIN__ 4.94065645841246544176568792868221372e-324L
-#define __LDBL_DIG__ 31
-#define __LDBL_EPSILON__ 4.94065645841246544176568792868221372e-324L
+#define __LDBL_DECIMAL_DIG__ 36
+#define __LDBL_DENORM_MIN__ 6.47517511943802511092443895822764655e-4966L
+#define __LDBL_DIG__ 33
+#define __LDBL_EPSILON__ 1.92592994438723585305597794258492732e-34L
 #define __LDBL_HAS_DENORM__ 1
 #define __LDBL_HAS_INFINITY__ 1
 #define __LDBL_HAS_QUIET_NAN__ 1
-#define __LDBL_IS_IEC_60559__ 0
-#define __LDBL_MANT_DIG__ 106
-#define __LDBL_MAX_10_EXP__ 308
-#define __LDBL_MAX__ 1.79769313486231580793728971405301199e+308L
-#define __LDBL_MAX_EXP__ 1024
-#define __LDBL_MIN_10_EXP__ (-291)
-#define __LDBL_MIN__ 2.00416836000897277799610805135016205e-292L
-#define __LDBL_MIN_EXP__ (-968)
-#define __LDBL_NORM_MAX__ 8.98846567431157953864652595394501288e+307L
+#define __LDBL_IS_IEC_60559__ 2
+#define __LDBL_MANT_DIG__ 113
+#define __LDBL_MAX_10_EXP__ 4932
+#define __LDBL_MAX__ 1.18973149535723176508575932662800702e+4932L
+#define __LDBL_MAX_EXP__ 16384
+#define __LDBL_MIN_10_EXP__ (-4931)
+#define __LDBL_MIN__ 3.36210314311209350626267781732175260e-4932L
+#define __LDBL_MIN_EXP__ (-16381)
+#define __LDBL_NORM_MAX__ 1.18973149535723176508575932662800702e+4932L
```
so maybe leave it at 11 after all?

and I'm actually testing with 11, not 12.
12 does have some fixes in that area,  but pretty obscure ones.

and I'm not sure about llvm/clang yet too.
my guess it's not important on gcc/glibc system as gcc/glibc/libstdc++ decide that and we don't have musl/clang for ppc64 (at least yet)

**Comment by thesamesam** [comment]
The first bug you filed is an example of something broken in 11 but fixed in 12, I think? (https://bugzilla.redhat.com/show_bug.cgi?id=2045186)

What GCC version are you specifically testing with? (Not just "11")?

**Comment by gyakovlev** [comment]
11.3.0

abseil bug in fedora is FTBS, for me it's just test failure, so a bit different.

anyway, I did some research and 12.x looks like better supported overall so I'll do that.

point about masking anything below that still remains though... we have to guarantee default `-mabi=ieeelongdouble system-wide`

**Comment by thesamesam** [comment]
> 11.3.0
> 
> abseil bug in fedora is FTBS, for me it's just test failure, so a bit different.

It goes on to discuss the test failure.

> 
> anyway, I did some research and 12.x looks like better supported overall so I'll do that.
> 

Agreed.

> point about masking anything below that still remains though... we have to guarantee default `-mabi=ieeelongdouble` system-wide

It'll be fine to mask older GCCs I think. GCC 12 will be fine to stable, the only problem is the PPC miscompilation for LLVM, which I think soap wants to look at on the weekend (https://bugs.gentoo.org/880677)

**Comment by gyakovlev** [comment]
rebasing this got old fast =)

File: profiles/arch/base/package.use.mask [hunk 2]
```diff
@@ -1,6 +1,11 @@
 # Copyright 1999-2022 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
+
```
Comment (thesamesam) [inline profiles/arch/base/package.use.mask:4]:
```suggestion
```

**Review by thesamesam — APPROVED** [review]

**Pull request closed without merging**